### ISSUE TITLE
Error parsing RestartPolicy from inspect response

### ISSUE THREAD

**Issue by mausch** [ahmetalpbalkan/Docker.DotNet#36]
Using Docker.DotNet 1.1.2.1 connecting to Docker 1.6.2. Calling `docker.Containers.InspectContainerAsync("something")` throws:

```
Error converting value "" to type 'Docker.DotNet.Models.RestartPolicyKind'. 
Path 'HostConfig.RestartPolicy.Name', line 1, position 1488.
```

Raw JSON response from inspection:

```json
[{
    "AppArmorProfile": "",
    "Args": [
        "elasticsearch",
        "--cluster.name=test-41e9c0ab-dbc5-41ca-8265-c4151746a4a9"
    ],
    "Config": {
        "AttachStderr": false,
        "AttachStdin": false,
        "AttachStdout": false,
        "Cmd": [
            "elasticsearch",
            "--cluster.name=test-41e9c0ab-dbc5-41ca-8265-c4151746a4a9"
        ],
        "CpuShares": 0,
        "Cpuset": "",
        "Domainname": "",
        "Entrypoint": [
            "/docker-entrypoint.sh"
        ],
        "Env": [
            "PATH=/usr/share/elasticsearch/bin:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin",
            "LANG=C.UTF-8",
            "JAVA_VERSION=8u45",
            "JAVA_DEBIAN_VERSION=8u45-b14-2~bpo8+2",
            "CA_CERTIFICATES_JAVA_VERSION=20140324",
            "ELASTICSEARCH_VERSION=1.7.1"
        ],
        "ExposedPorts": {
            "9200/tcp": {},
            "9300/tcp": {}
        },
        "Hostname": "dd145bb87386",
        "Image": "elasticsearch:1.7.1",
        "Labels": {},
        "MacAddress": "",
        "Memory": 0,
        "MemorySwap": 0,
        "NetworkDisabled": false,
        "OnBuild": null,
        "OpenStdin": false,
        "PortSpecs": null,
        "StdinOnce": false,
        "Tty": false,
        "User": "",
        "Volumes": {
            "/usr/share/elasticsearch/data": {}
        },
        "WorkingDir": ""
    },
    "Created": "2015-09-04T13:18:55.886520689Z",
    "Driver": "aufs",
    "ExecDriver": "native-0.2",
    "ExecIDs": null,
    "HostConfig": {
        "Binds": null,
        "CapAdd": null,
        "CapDrop": null,
        "CgroupParent": "",
        "ContainerIDFile": "",
        "CpuShares": 0,
        "CpusetCpus": "",
        "Devices": null,
        "Dns": null,
        "DnsSearch": null,
        "ExtraHosts": null,
        "IpcMode": "",
        "Links": null,
        "LogConfig": {
            "Config": null,
            "Type": "json-file"
        },
        "LxcConf": null,
        "Memory": 0,
        "MemorySwap": 0,
        "NetworkMode": "",
        "PidMode": "",
        "PortBindings": null,
        "Privileged": false,
        "PublishAllPorts": true,
        "ReadonlyRootfs": false,
        "RestartPolicy": {
            "MaximumRetryCount": 0,
            "Name": ""
        },
        "SecurityOpt": null,
        "Ulimits": null,
        "VolumesFrom": null
    },
    "HostnamePath": "/mnt/b53c2b12-0aed-4083-aefc-7c4cb0046360/docker/containers/dd145bb873863187e1d3e508e475d36a65dddc8fb1fb8f4c74818e066169ff5d/hostname",
    "HostsPath": "/mnt/b53c2b12-0aed-4083-aefc-7c4cb0046360/docker/containers/dd145bb873863187e1d3e508e475d36a65dddc8fb1fb8f4c74818e066169ff5d/hosts",
    "Id": "dd145bb873863187e1d3e508e475d36a65dddc8fb1fb8f4c74818e066169ff5d",
    "Image": "bd58cf964c1b13de2a89cd7b6b14a15fb49f2780b07d136a7cd6bb7a93611bf4",
    "LogPath": "/mnt/b53c2b12-0aed-4083-aefc-7c4cb0046360/docker/containers/dd145bb873863187e1d3e508e475d36a65dddc8fb1fb8f4c74818e066169ff5d/dd145bb873863187e1d3e508e475d36a65dddc8fb1fb8f4c74818e066169ff5d-json.log",
    "MountLabel": "",
    "Name": "/desperate_perlman",
    "NetworkSettings": {
        "Bridge": "docker0",
        "Gateway": "172.17.42.1",
        "GlobalIPv6Address": "",
        "GlobalIPv6PrefixLen": 0,
        "IPAddress": "172.17.0.8",
        "IPPrefixLen": 16,
        "IPv6Gateway": "",
        "LinkLocalIPv6Address": "fe80::42:acff:fe11:8",
        "LinkLocalIPv6PrefixLen": 64,
        "MacAddress": "02:42:ac:11:00:08",
        "PortMapping": null,
        "Ports": {
            "9200/tcp": [
                {
                    "HostIp": "0.0.0.0",
                    "HostPort": "32780"
                }
            ],
            "9300/tcp": [
                {
                    "HostIp": "0.0.0.0",
                    "HostPort": "32781"
                }
            ]
        }
    },
    "Path": "/docker-entrypoint.sh",
    "ProcessLabel": "",
    "ResolvConfPath": "/mnt/b53c2b12-0aed-4083-aefc-7c4cb0046360/docker/containers/dd145bb873863187e1d3e508e475d36a65dddc8fb1fb8f4c74818e066169ff5d/resolv.conf",
    "RestartCount": 0,
    "State": {
        "Dead": false,
        "Error": "",
        "ExitCode": 0,
        "FinishedAt": "0001-01-01T00:00:00Z",
        "OOMKilled": false,
        "Paused": false,
        "Pid": 5361,
        "Restarting": false,
        "Running": true,
        "StartedAt": "2015-09-04T13:18:56.21906321Z"
    },
    "Volumes": {
        "/usr/share/elasticsearch/data": "/mnt/b53c2b12-0aed-4083-aefc-7c4cb0046360/docker/vfs/dir/66947008e712e3dc0eb153f717a16c35685d1962bb5f74db0df0dc91af2c54b0"
    },
    "VolumesRW": {
        "/usr/share/elasticsearch/data": true
    }
}
]
```

The response parser should use ["no" as the default value](https://docs.docker.com/reference/run/#restart-policies-restart). 
By the way that doesn't seem to be present in `Docker.DotNet.Models.RestartPolicyKind` at the moment?

**Comment by ahmetalpbalkan** [comment]
I am not sure "no" in the CLI actually maps to API response. The empty value used to mean "no" and that's why we don't have it. 

Do you think adding "no" would fix it?

**Comment by mausch** [comment]
IMHO "no" should be added because it's one of the documented restart policies. If I had a function building container parameters it would be much easier to return "no" instead of having to figure out whether to add the policy parameter or not (is this supported for creating containers in this client lib? I haven't checked TBH)

Of course I don't like Docker's empty response here... it should be explicit, but that's probably something to talk about with the Docker maintainers. Client libraries should be able to handle this odd behavior (with a custom deserializer or whatever).

Cheers

**Comment by ahmetalpbalkan** [comment]
> IMHO "no" should be added because it's one of the documented restart policies

@mausch  Can you please elaborate? I can't seem to find it in the docs. 

> 
![image](https://cloud.githubusercontent.com/assets/159209/9707632/f73b9f66-54c2-11e5-9fce-13cbcd84dbf2.png)

If you're referring to `docker run --restart=no`, that doesn't mean it means the same thing in the API. I have not seen an indication of "no" being a valid input to the API.

**Comment by mausch** [comment]
Ahh, sorry, I understand now. Yes, I was just using the CLI as reference, my mistake.

Then it sounds like the right thing to do here is to parse that response as a null RestartPolicy in the HostConfig instance, what do you think?

**Comment by ahmetalpbalkan** [comment]
@mausch yes sounds like parsing into null is a good choice here.

**Comment by charoon** [comment]
with docker api 1.18 I get 
> "Error converting value \"no\" to type 'Docker.DotNet.Models.RestartPolicyKind'"

in json 

> "RestartPolicy":{"MaximumRetryCount":0,"Name":"no"}

IMHO parsing restart policy 'no' into null is not good choise.
@ahmetalpbalkan this part of doc is more accurately for container create request, not for inspect response and in response, I see 'no'

**Comment by ahmetalpbalkan** [comment]
@charoon on which request do you get this error? Is it for creating containers? If so, you should not specify "no", your restart policy should just be null ("no" is a word that exists in docker client, not the docker remote API).

**Comment by charoon** [comment]
@ahmetalpbalkan error is from inspect response. I get 'no' in inspect json response, not '' as in first post

**Comment by ahmetalpbalkan** [comment]
@charoon I see, in my tests API version 1.19 definitely returns

    "RestartPolicy":{"MaximumRetryCount":0,"Name":"no"}

even though no `--restart` is specified. @mausch what's the Docker Server API version you are using (see `docker version`)?

Looks like we need to add a simple enum to `Docker.DotNet.Models.RestartPolicyKind`, @charoon want to send a PR?

**Comment by mausch** [comment]
Server API version: 1.18

The docs for 1.20 show an empty name in an example in https://docs.docker.com/reference/api/docker_remote_api_v1.20/#create-a-container , although that could be just outdated documentation.

**Comment by ahmetalpbalkan** [comment]
@mausch empty string actually might be still supported in the `CreateContainer` operation... `InspectContainer` on the other hand is a different story.

Looks like 1.18+ docker returns 

             "RestartPolicy": { "Name": "no", "MaximumRetryCount": 0 },

Instead of empty string in the Inspect/Get responses. We need to edit `Docker.DotNet.Models.RestartPolicyKind`... @mausch want to send a pull request?

**Comment by adebasi** [comment]
I have the same problem. Using the CLI I receive an emptry string or "no" as RestartPolicy-Name - depends on the container.

Client:
 Version:      1.8.3
 API version:  1.20

Server:
 Version:      1.8.3
 API version:  1.20

What's the status for this issue?

**Comment by mausch** [comment]
@adebasi Sorry, I haven't looked further into this as I was only evaluating the possibility of scripting Docker in .NET... but it seems that I won't get enough buy-in from the rest of the team for this to happen.

**Comment by adebasi** [comment]
Thanks for the update. For now I set the RestartPolicyKind to OnFailure during container creation to avoid the bug. This works for me.

**Comment by ahmetalpbalkan** [comment]
@adebasi could it be because some of your containers are created before docker-1.9 and some after upgrading to 1.9 and that's why you're receving mixed `"no"`/`""` responses about RestartPolicy.Name in `docker inspect`?

**Comment by adebasi** [comment]
@ahmetalpbalkan I figured out, why I have different responses. I created new containers to make sure it's not a version issue. Docker versions are still the same like I posted yesterday.

One container was created with your framework like this and returned emtpy string as RestartPolicy-Name.
```await client.Containers.StartContainerAsync(containerId,new HostConfig{});```
The other container was created via CLI and returned "no" as RestartPolicy-Name.
```docker run -d <imagename>```
